File: 프로그래머스/lv3/43164. 여행경로/여행경로.py
from collections import deque

def solution(tickets):
    tickets.sort()
    visit = [False]*len(tickets)

    def dfs(src, path):
        if len(path) == len(tickets)+1:
            answer.append(path)
            return

        for i in range(len(tickets)):
            if src == tickets[i][0] and not visit[i]:
                visit[i] = True
                dfs(tickets[i][1], path+[tickets[i][1]])
                visit[i] = False #백트래킹

    answer = []
    dfs('ICN', ['ICN'])
    return answer[0]

File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

vector<vector<string>> path_results;
int N;
vector<bool> visit;

void dfs(string cur, vector<string> path, vector<vector<string>> &tickets)
{
    if (path.size() == N+1)
    {
        path_results.push_back(path);
        return;
    }
    else
    {
        for (int i = 0; i < N; ++i)
        {
            if (!visit[i] && tickets[i][0] == cur)
            {
                visit[i] = true; path.push_back(tickets[i][1]);
                dfs(tickets[i][1], path, tickets);
                visit[i] = false; path.pop_back();
            }
        }
    }
}

vector<string> solution(vector<vector<string>> tickets) {
    
    N = tickets.size();

    sort(tickets.begin(), tickets.end());
    visit.resize(N, false);
    dfs("ICN", {"ICN"}, tickets);
    
    return path_results[0];
}